src/ddspdrum/module.py

```diff
@@ -0,0 +1,239 @@
+        TODO: One thing I worry about with all these casts to
+        ints back and forth between sample and control rate is
+        off-by-one errors.
+        I'm beginning to believe we should standardize as much
+        as possible on nsamples in most places instead of
+        float duration in seconds, and just convert to ncontrol
+        when needed..
+        """
+        # Right now it appears that all signals are 1d, but later
+        # we'll probably convert things to 2d: instance x signal
+        assert control.ndim == 1
+        nsamples = int(round(len(control) * self.sample_rate / self.control_rate))
+        return resample(control, nsamples)
+
+    def fix_length(self, signal: np.array, length: int) -> np.array:
+        # Right now it appears that all signals are 1d, but later
+        # we'll probably convert things to 2d: instance x signal
+        assert signal.ndim == 1
+        if len(signal) < length:
+            signal = np.pad(signal, [0, length - len(signal)])
+        elif len(signal) > length:
+            signal = signal[:length]
+        assert signal.shape == (length,)
+        return signal
+
+
+class ADSR(SynthModule):
+    """
+    Envelope class for building a control rate ADSR signal
+    """
+
+    def __init__(
+        self,
+        a: float = 0.25,
+        d: float = 0.25,
+        s: float = 0.5,
+        r: float = 0.5,
+        alpha: float = 3.0,
+        sample_rate: int = SAMPLE_RATE,
+        control_rate: int = CONTROL_RATE,
+    ):
+        """
+        Parameters
+        ----------
+        a                   :   attack time (sec), >= 0
+        d                   :   decay time (sec) in [0, 1]
+        s                   :   sustain amplitude between 0-1
+        r                   :   release time (sec), >= 0
+        alpha               :   envelope curve, >= 0. 1 is linear, >1 is exponential.
+        """
+        super().__init__(sample_rate=sample_rate, control_rate=control_rate)
+        assert alpha >= 0
+        self.alpha = alpha
+
+        assert s >= 0 and s <= 1
+        self.s = s
+
+        assert a >= 0
+        self.a = a
+
+        # TODO: WHY? does it have to have an upper bound of 1?
+        assert d >= 0 and d <= 1
+        self.d = d
+
+        assert r >= 0
+        self.r = r
+
+    def __call__(self, duration):
+        """
+        Play, for some duration in seconds.
+        """
+        assert duration >= 0
+        return np.concatenate(
+            (
+                # Is this the DSP-way of thinking about it, (note on and note off)
+                # or can we change this just to self.attack, self.decay, etc.
+                self.note_on,
+                # This makes me believe that duration should maybe be a class value
+                # defined in __init__. How common is it to create a module and
+                # with fixed settings and use it several times with different
+                # duration?
+                self.sustain(duration),
```

Comment:
So one question we will need to solve is:
* what parameters go into `__init__`
* what parameters go into `forward`?

I am not sure whether this sustain duration should go into `forward` or `__init__`. As a rule of thumb, if it's the output from another module, then it goes into `forward` and otherwise it's a setting that goes in `__init__`.

Reply:
Makes sense to me to set parameters in the constructor and then the output from another module gets passed into forward. It would be nice to have a way to set parameters after instantiation, I can see that be good for use after training, but we can cross that bridge later.

Reply:
At the moment, the envelope sets the global duration; that sets the length of the control signals, which ultimately sets the length of the audio signals. 

**Amp**-envelope parameters will determine the note-length (e.g., a release time of 0.5 adds 0.5 secs). Also the sustain amount.

Mind you, a frequency **cutoff**-envelope or a **pitch**-envelope can be shorter than the amp envelope, and just be padded with zeroes after to match the amp envelope. So we have to be careful about this. But when the amp envelope ends, the signal ends.

Maybe best to instantiate a signal with **amp** envelope ADR params plus sustain_duration. Then we have a fixed signal length and can allocate memory accordingly. What's tricky is that the amp envelope is not always the first step of the signal chain, in fact it's often the last.

Reply:
@maxsolomonhenry I don't exactly understand the concerns. Can you explain?

Which of these parameters is influenced by the output from another module which is input to this module? If we receive a freq cutoff or pitch envelope that is short then the rest of the parameters don't matter?

"Then we have a fixed signal length and can allocate memory accordingly." Does it make sense that the signal length should be---i don't want to say global---but at least something you define in advance before running the synth? Like sample rate and control rate?

Still don't fully grok this yet.

Reply:
Here's the idea. Let's assume: sample_rate = 44100, and control_rate= 1000, sustain_duration = 0.1s

Let's say you want a signal with a quick, snappy pitch attack. But you want the note to ring for a long time. 
You make an envelope for pitch--> 
```
pitch_env = ADSR(a=0.01, d=0.01,s=0,r=0)
```
This will be (0.01 + 0.01 + 0.1)*1000 = 120 samples long. But you want the sound to ring out, so you have an amp envelope with different parameters -->
```
amp_env = ADSR(a=0.01, d=0.25, s=0.75, r=0.5)
```
This will give an envelope that is (0.01 + 0.25 + 0.1 + 0.75) * 1000 = 1110 samples long.

This means we want a signal that is 1110 * 44100/1000 = 48951 samples at the audio rate.

Reply:
@jorshi "It would be nice to have a way to set parameters after instantiation, I can see that be good for use after training, but we can cross that bridge later."

You just modify the member variable of the class, and then run the module again.

Reply:
@maxsolomonhenry So is what you're saying that we might have shorter pitch envelope and a shorter signal earlier in the chain, but the later modules might require longer signals and thus padding the earlier signal with zeros? I guess we can either a) zero pad or b) determine the final signal length at the beginning, with a parameter like `sample_rate` but instead `sample_duration`. a) sounds easier to start.